shadowsocks/stream.go

```diff
@@ -165,76 +166,91 @@ func (sr *shadowsocksReader) init() (err error) {
 	return nil
 }
 
-// ReadBlock reads and decrypts a single signed block of ciphertext.
-// The block will match the given decryptedBlockSize.
-// The returned slice is only valid until the next Read call.
-func (sr *shadowsocksReader) readBlock(decryptedBlockSize int) ([]byte, error) {
-	if err := sr.init(); err != nil {
-		return nil, err
-	}
-	cipherBlockSize := decryptedBlockSize + sr.aead.Overhead()
-	if cipherBlockSize > cap(sr.buf) {
-		return nil, io.ErrShortBuffer
-	}
-	buf := sr.buf[:cipherBlockSize]
+// readBlock reads and decrypts a single signed block of ciphertext.
+// The block must exactly match the size of `buf`, including overhead.
+// Returns an error only if the block could not be read.
+func (sr *shadowsocksReader) readBlock(buf []byte) error {
```

Comment:
Maybe `readCipherBlock` or `readEncryptedBlock`? I see you are using the original name, but I think it will make it clearer that this is reading an AEAD block.

Reply:
Changed to `readMessage`